File: php/74-search-a-2d-matrix.php
<?php
class Solution {

    /**
     * @param Integer[][] $matrix
     * @param Integer $target
     * @return Boolean
     */
    function searchMatrix($matrix, $target) {
        $row = count($matrix);
        $col = count($matrix[0]);

        for ($i = 0; $i < $row; $i++) {
            $ele = $matrix[$i][$col - 1];
            if ($ele == $target) {
                return true;
            }
            if ($target < $ele) {
                for ($j = 0; $j < $col; $j++) {
                    if ($matrix[$i][$j] == $target) {
                        return true;
                    }
                }
                return false;
            }
        }
        return false;
    }
}

File: go/74-search-a-2d-matrix.go
package main

func searchMatrix(matrix [][]int, target int) bool {
    row := len(matrix)
    if row == 0 {
        return false
    }
    col := len(matrix[0])
    right := row * col - 1
    left := 0
    return findRow(matrix, target, left, right)
}
func findRow(matrix [][]int, target, left, right int) bool {
    if left > right {
        return false
    }
    var row, col, mid int
    mid = (left + right) / 2
    row = mid / len(matrix[0])
    col = mid - row * len(matrix[0])
    if matrix[row][col] == target {
        return true
    }
    if matrix[row][col] < target {
        return findRow(matrix, target, mid + 1, right)
    } else {
        return findRow(matrix, target, left, mid - 1)
    }
}
